Classes/common/HTTP/CDTSessionCookieInterceptor.m

```diff
@@ -0,0 +1,147 @@
+    }
+
+    return context;
+}
+
+- (CDTHTTPInterceptorContext *)interceptResponseInContext:(CDTHTTPInterceptorContext *)context
+{
+    // check the response code.
+    if (context.response.statusCode == 401) {
+        // reset the cookie, let the request interceptor handle getting a new cookie.
+        self.cookie = nil;
+        context.shouldRetry = YES;
+    }
+
+    return context;
+}
+
+- (nullable NSString *)newCookieForRemote:(NSURL *)url
+{
+    NSString *scheme = [url scheme];
+    NSNumber *port = [url port];
+    if (!port) {
+        if ([scheme isEqualToString:@"https"]) {
+            port = @(443);
+        } else {
+            port = @(80);
+        }
+    }
+
+    NSURL *sessionUrl = [NSURL
+        URLWithString:[NSString stringWithFormat:@"%@://%@:%@/_session", scheme, [url host], port]];
+
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:sessionUrl];
+    request.HTTPBody = [self.cookieRequestBody dataUsingEncoding:NSUTF8StringEncoding];
+    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
+
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+    __block NSString *cookie = nil;
+    NSURLSessionDataTask *task = [self.urlSession
+        dataTaskWithRequest:request
+          completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+            // all parameters to block are nullable, annotations removed
+            // to provide compatbility with xcode 6.4 and xcode 7
+            NSHTTPURLResponse *httpResp = (NSHTTPURLResponse *)response;
+            // handle response being nil
+            if ([httpResp statusCode] / 100 == 2) {
+                // we have the cookie. maybe
+                if (data && [self hasSessionStarted:data]) {
+                    // get the cookie from the header
+                    NSString *cookieHeader = [[httpResp allHeaderFields] valueForKey:@"Set-Cookie"];
+
+                    cookie = [cookieHeader componentsSeparatedByString:@";"][0];
+
+                } else {
+                    cookie = nil;
+                }
+            } else if ([httpResp statusCode] == 401) {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"Credentials are incorrect, cookie "
+                                                        @"authentication will not be attempted "
+                                                        @"again by this interceptor object");
+                self.shouldMakeCookieRequest = NO;
+            } else if ([httpResp statusCode] / 100 == 5) {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT,
+                            @"Failed to get cookie from the server, response code was %ld cookie "
+                            @"authentication will not be attempted again by this interceptor "
+                            @"object",
+                            (long)[httpResp statusCode]);
+                self.shouldMakeCookieRequest = NO;
```

Comment:
I don't think this is valid for `500` status codes. `500` errors tend to be transient.